libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -129,7 +129,7 @@ string IRGenerator::generate(
 	vector<string> constructorParams;
 	if (constructor && !constructor->parameters().empty())
 	{
-		for (size_t i = 0; i < constructor->parameters().size(); ++i)
+		for (size_t i = 0; i < CompilerUtils::sizeOnStack(constructor->parameters()); ++i)
```

Comment:
Note to self: Before this PR, a new yul variable was created per constructor parameter. This is okay for all types except perhaps function, which presumably requires more than one variable to be identified (contract address plus function selector).

This meant that during IR generation for constructors, one parameter was placed in the constructor definition but a function call actually passed two arguments.

Reply:
@hrkrshnn Can you please point to where function calls are happening? Do call sites use `CompilerUtils::sizeOnStack`?

Reply:
From the issue, the problem is

```
        let _1 := copy_arguments_for_constructor_10_object_D_11()
        constructor_D_11(_1)
```

but the constructor is actually generated properly

```
        function constructor_D_11(vloc__6_address, vloc__6_functionSelector) {

        }
```

So the functions `copy_arguments...` and the constructor call needed to have the correct amount of "stack slots". For function types, there are 2 slots. Previously the number of slots needed were calculated as just the number of parameter types, this calculation only works for single slotted types.

Does that help?

Reply:
> Does that help?

Yes, definitely. Thank you for clarifying. I'm just wondering if there are other places where this fix may be needed. Is function type the only type which requires more than one slot? Also, what about functions that accept function types as arguments? Do call sites for these functions copy the correct number of arguments?

Reply:
Functions seem to work fine. For example

```
contract D {
    function f(
	function() internal returns (uint) input1,
	function() internal returns (uint) input2
	) internal returns (function() internal returns (uint)) {
	return input1;
    }
    function g() external returns (uint) {
	return f(h, i)();
    }
    function h() internal returns (uint) {
	return 42;
    }
    function i() internal returns (uint) {
	return 1337;
    }
}

// ====
// compileViaYul: also
// ----
// g() -> 0x2a
```

Reply:
> Is function type the only type which requires more than one slot

You can check the `makeStackItems` from `Types.cpp`. Apparently dynamic calldata arrays take two stack slots. But, you can't have calldata parameters in constructors. So I guess for constructor, function types are the only type that needs two stack slots.

Reply:
Functions are fine, it was just constructor that had the issue. @bshastry